x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizer.java

```diff
@@ -89,62 +80,26 @@ static class ProjectAwayColumns extends Rule<PhysicalPlan, PhysicalPlan> {
 
         @Override
         public PhysicalPlan apply(PhysicalPlan plan) {
-            var projectAll = new Holder<>(TRUE);
-            var keepCollecting = new Holder<>(TRUE);
-            var attributes = new AttributeSet();
-            var aliases = new AttributeMap<Expression>();
+            Holder<Boolean> keepTraversing = new Holder<>(TRUE);
+            // Invariant: if we add a projection with these attributes after the current plan node, the plan remains valid
+            // and the overall output will not change.
+            Holder<AttributeSet> requiredAttributes = new Holder<>(plan.outputSet());
 
-            return plan.transformDown(UnaryExec.class, p -> {
-                // no need for project all
-                if (p instanceof ProjectExec || p instanceof AggregateExec) {
-                    projectAll.set(FALSE);
+            // This will require updating should we choose to have non-unary execution plans in the future.
+            return plan.transformDown(UnaryExec.class, currentPlanNode -> {
+                if (keepTraversing.get() == false) {
+                    return currentPlanNode;
                 }
-                if (keepCollecting.get()) {
-                    p.forEachExpression(NamedExpression.class, ne -> {
-                        var attr = ne.toAttribute();
-                        // filter out attributes declared as aliases before
-                        if (ne instanceof Alias as) {
-                            aliases.put(attr, as.child());
-                            attributes.remove(attr);
-                        } else {
-                            // skip synthetically added attributes (the ones from AVG), see LogicalPlanOptimizer.SubstituteSurrogates
-                            if (attr.synthetic() == false && aliases.containsKey(attr) == false) {
```

Comment:
Synthetic attributes were skipped because we had to solve an NPE; IMHO this was not a correct long term fix, as it made heavy assumptions on where synthetic attributes are used.